src/drivers/mediatek/mt8195/interrupt.c

```diff
@@ -0,0 +1,559 @@
+		if ((1 << (i - 32)) & mask) {
+			if (ext_isr[i].handler) {
+				currnt_handled_irq = i;
+				ext_isr[i].handler();
+			} else {
+				tr_err(&int_tr, "EXT_DOMAIN2_MASK no handler irq: %d\n", i);
+				mt_ext_mask_module(i);
+			}
+		}
+	}
+}
+
+static void mt_ext_init(void)
+{
+	uint32_t level;
+
+	/* external domain1 irq init */
+	level = GET_INTLEVEL(L1_INT_IRQ_B);
+	_xtos_set_interrupt_handler_arg(level, mt_ext_handle_domain1_irq, (void *)level);
+	_xtos_ints_on(1 << level);
+
+	/* external domain2 irq init */
+	level = GET_INTLEVEL(L23_INT_IRQ_B);
+	_xtos_set_interrupt_handler_arg(level, mt_ext_handle_domain2_irq, (void *)level);
+	_xtos_ints_on(1 << level);
+
+}
+
+uint64_t get_max_cs_duration_time(void)
+{
+	return max_cs_duration_time;
+}
+
+uint64_t get_max_cs_limit(void)
+{
+	return max_cs_limit;
+}
+
+uint32_t read_taken_INT(void)
+{
+	return currnt_handled_irq;
+}
+
+void set_irq_limit(uint32_t irq_limit_time)
+{
+	if (irq_limit_time == 0) {
+		max_irq_limit = 0;
+	} else {
+		max_irq_limit = IRQ_DURATION_LIMIT;
+	}
+}
+
+static inline void irq_handler(void *data, uint32_t line_index)
+{
+}
+
+/** register a irq handler
+ *
+ *  @param irq interrupt number
+ *  @param irq interrupt handler
+ *  @param irq interrupt name
+ *
+ *  @returns
+ *    no return
+ */
+void request_irq(uint32_t irq, irq_handler_t handler, const char *name)
+{
+	uint32_t level, in_irq;
+
+	level = GET_INTLEVEL(irq);
+	in_irq = GET_INTERRUPT_ID(irq);
+
+	if (level > MAX_INT_LEVEL) {
+		tr_err(&int_tr, "ERR Int level %u\n\r", level);
+		return;
+	}
+
+	switch (level) {
+	case 1:
+	case 9:
+	case 23:
+	case 24:
+		mt_ext_register_isr(in_irq, handler, name);
+		mt_ext_unmask_module(in_irq);
+		break;
+
+	default:
+		if (in_irq > MAX_IRQ_NUM ||	\
+			(irq_desc[level].handler)) {
+			tr_err(&int_tr, "ERR Int irq %u\n\r", in_irq);
+			return;
+		}
+		_xtos_ints_off(1 << level);
+		mt_ext_mask_module(in_irq);
+
+		irq_desc[level].name = name;
+		irq_desc[level].handler = handler;
+		irq_desc[level].init_count++;
+
+		mt_ext_unmask_module(in_irq);
+		_xtos_ints_on(1 << level);
+		break;
+	}
+}
+
+/** free a irq handler
+ *  @param irq interrupt number
+ *
+ *  @returns
+ *    no return
+ */
+void free_irq(uint32_t irq)
+{
+	uint32_t level, in_irq;
+
+	level = GET_INTLEVEL(irq);
+	in_irq = GET_INTERRUPT_ID(irq);
+	if (level > MAX_INT_LEVEL) {
+		tr_err(&int_tr, "ERR Int level %u\n\r", level);
+		return;
+	}
+
+	switch (level) {
+	case 1:
+	case 9:
+	case 23:
+	case 24:
+		mt_ext_deregister_isr(in_irq);
+		mt_ext_mask_module(in_irq);
+		break;
+
+	default:
+		if (in_irq > MAX_IRQ_NUM) {
+			tr_err(&int_tr, "ERR Int irq %u\n\r", in_irq);
+			return;
+		}
+		_xtos_ints_off(1 << level);
+		mt_ext_mask_module(in_irq);
+		irq_desc[level].name = NULL;
+		irq_desc[level].handler = NULL;
+		break;
+	}
+}
+
+#define DEFINE_IRQ_HANDLER(n) \
+	static inline void irqstr_irqhandler_##n(void *arg) \
+	{ \
+		irq_handler(arg, n); \
+	}
```

Comment:
Does this macro even make sense for you? I've written it specifically for IRQ_STEER which is NXP specific, and redirects 512 global interrupts into 8 DSP interrupts.